File: backend/schemas/briefing.py
"""
schemas/briefing.py — Pydantic models for the full briefing payload
and history list items.
"""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field

from schemas.calendar import CalendarEvent, CalendarProposal, FreeSlot
from schemas.email import ClassifiedEmail
from schemas.tasks import ClassifiedTask


class CrossReference(BaseModel):
    """Links an email, calendar event, and/or task that are related to each other."""

    email_id: str | None = None
    event_id: str | None = None
    task_id: str | None = None
    relationship_description: str = Field(
        ...,
        description="Human-readable explanation of why these items are related.",
    )


class Briefing(BaseModel):
    """The full morning briefing payload returned to the frontend."""

    # Metadata
    id: str = Field(default="", description="Set by stage6_persist after DB insert.")
    user_id: str
    generated_at: datetime
    date: date

    # AI-generated summary
    focus_suggestion: str = Field(
        ...,
        description="One-paragraph AI recommendation for how to focus the day.",
    )

    # High-level stats
    urgent_email_count: int = 0
    meeting_count: int = 0
    overdue_task_count: int = 0
    calendar_conflict_flag: bool = False

    # Detailed data
    emails: list[ClassifiedEmail] = Field(default_factory=list)
    events: list[CalendarEvent] = Field(default_factory=list)
    free_slots: list[FreeSlot] = Field(default_factory=list)
    tasks: list[ClassifiedTask] = Field(default_factory=list)

    # AI enrichments
    cross_references: list[CrossReference] = Field(default_factory=list)
    proposals: list[CalendarProposal] = Field(default_factory=list)

    # Degraded / partial state
    partial: bool = Field(
        default=False,
        description="True if one or more data sources were unavailable.",
    )
    unavailable_sources: list[str] = Field(
        default_factory=list,
        description="Names of sources that failed (e.g. ['gmail', 'notion']).",
    )


class BriefingHistoryItem(BaseModel):
    """Lightweight summary for the briefing history list view."""

    id: str
    date: date
    generated_at: datetime
    focus_suggestion: str | None
    urgent_email_count: int
    meeting_count: int
    overdue_task_count: int
    calendar_conflict_flag: bool
    partial: bool

    model_config = {"from_attributes": True}

File: frontend/src/types/briefing.ts
/** Urgency classification for an email. */
export type EmailClassification = 'urgent' | 'can_wait' | 'fyi'

/** A Gmail message enriched with AI classification. */
export interface ClassifiedEmail {
  id: string
  subject: string
  sender: string
  sender_email: string
  snippet: string
  received_at: string // ISO 8601
  classification: EmailClassification
  classification_reason: string
  thread_id: string
}

/** A Google Calendar event with conflict/back-to-back metadata. */
export interface CalendarEvent {
  id: string
  title: string
  start: string // ISO 8601
  end: string   // ISO 8601
  attendees: string[]
  description: string | null
  location: string | null
  conflict_flag: boolean
  back_to_back_flag: boolean
  conflict_with_id: string | null
}

/** A contiguous free time window on the calendar. */
export interface FreeSlot {
  start: string  // ISO 8601
  end: string    // ISO 8601
  duration_minutes: number
}

/** Due-date classification for a Notion task. */
export type TaskClassification = 'overdue' | 'due_today' | 'due_later'

/** A Notion task enriched with a deterministic due-date classification. */
export interface ClassifiedTask {
  id: string
  title: string
  due_date: string | null // ISO date YYYY-MM-DD
  classification: TaskClassification
  project: string | null
  url: string
}

/** A semantic link between an email, event, and/or task. */
export interface CrossReference {
  email_id: string | null
  event_id: string | null
  task_id: string | null
  relationship_description: string
}

/** An AI-generated suggestion to schedule a task in a free slot. */
export interface CalendarProposal {
  task_id: string
  task_title: string
  proposed_start: string // ISO 8601
  proposed_end: string   // ISO 8601
  rationale: string
}

/** The full morning briefing payload. */
export interface Briefing {
  user_id: string
  generated_at: string // ISO 8601
  date: string         // YYYY-MM-DD
  focus_suggestion: string

  // Stats
  urgent_email_count: number
  meeting_count: number
  overdue_task_count: number
  calendar_conflict_flag: boolean

  // Detailed data
  emails: ClassifiedEmail[]
  events: CalendarEvent[]
  free_slots: FreeSlot[]
  tasks: ClassifiedTask[]

  // Enrichments
  cross_references: CrossReference[]
  proposals: CalendarProposal[]

  // Degraded state
  partial: boolean
  unavailable_sources: string[]
}

/** Lightweight summary for the briefing history list. */
export interface BriefingHistoryItem {
  id: string
  date: string         // YYYY-MM-DD
  generated_at: string // ISO 8601
  focus_suggestion: string | null
  urgent_email_count: number
  meeting_count: number
  overdue_task_count: number
  calendar_conflict_flag: boolean
  partial: boolean
}
